rclcpp/src/rclcpp/context.cpp

```diff
@@ -341,20 +348,38 @@ Context::shutdown(const std::string & reason)
   return true;
 }
 
-rclcpp::Context::OnShutdownCallback
-Context::on_shutdown(OnShutdownCallback callback)
+rclcpp::OnShutdownCallbackHandle::SharedPtr
+Context::on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback)
 {
-  on_shutdown_callbacks_.push_back(callback);
-  return callback;
+  std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
+  auto handle = std::make_shared<OnShutdownCallbackHandle>();
```

Comment:
If we keep the idea of returning an scoped handle, this could have a custom deleter that automatically calls `remove_on_shutdown_callback()` (leveraging this->weak_from_this()).